### PR TITLE
Allows binding bean of inaccessible type (Option 2)

### LINKED ISSUE #242: @BindBean is incompatible with objects of non-public types
If `@BindBean` is specified with a type (let's call it `Foo`) that is a `public` `class` or `interface`, but is passed an object whose type (`ImplementsFoo`) is defined as a non-public `class`, the result is an exception in `BindBeanFactory`:

  `java.lang.IllegalAccessException: ... can not access a member of class`

This is because an attempt is made to invoke a method of a non-public class.  Note this exception is generated even though calling this method would generally be allowed as a function of polymorphism.

My take is that `BindBeanFactory` should use `setAccessible(true)` and allow the method call `readMethod.invoke(arg)` if the base type's method is public regardless whether the provided object's access modifier is public of not.

References:
1) [Relevant code](https://github.com/jdbi/jdbi/blob/e206c5681a68a5887beaecd71781dd8584b54332/src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java#L48) in `BindBeanFactory`.
2) [A more in depth discussion with examples](http://stackoverflow.com/questions/34009266/using-jdbi-bindbean-with-autovalue?noredirect=1#comment55774666_34009266) of how I discovered this issue while attempting to use `JDBI` with `AutoValue` on StackOverflow.

### PR DESCRIPTION
**Pull request by gsmith85**
Changes the BinderFactory interface so that the type token of the
annotated parameter is accepted as a parameter on #build.

Changes BindBeanFactory to get the value of its argument using the
Method of the annotated parameter's type instead of the argument's type.
This allows for BindBeanFactory to access methods on args that are of
types that are inaccessible to the SqlObject package.  This changes the
behavior of the BindBeanFactory to behave in a way that is much more
consistent with Java's polymorphic/access mechanism.

Using BindBean with an iterable of a raw type will result in using
Object.  This may be considered a limiting factor or a feature.

Fixes #242

### CODE REVIEW

**Comment by gsmith85** [comment]
Consider this to be Option 1 of 2 (see #243) options to address the related issue.

**Comment by qualidafial** [comment]
This option is a non-starter for v2, since it modifies the parameters of an existing interface, which breaks backward compatibility with existing 3rd party binders.

For v3 however (which is in flight in the jdbi3 branch), this will probably be the preferred option going forward.

**Comment by qualidafial** [comment]
Let's wait until #243 is merge before deciding how this will look in v3.

**Comment by gsmith85** [comment]
The other strategy that's similar to this approach but *is* backwards compatible (in the sense we're not touching existing interfaces) would be adding a `ParameterTypeDependentBinder`, `ParameterTypeDependentBindingAnnotation` and then in `CustomizingStatementHandler` checking methods for this annotation as well as `BindingAnnotation`.  In the case of `ParameterTypeDependentBindingAnnotation`, we'd pass in the parameter type.

The downside of this is a bifurcated code path for creating binders and the need to add a couple additional types.  The upside is obviously backwards compatibility and allowing for developers to create their own parameter-type dependent bindings if they desired (you could actually add support for raw types for example if you wanted to or allow for duck-typing).

Using the type of the annotated parameter seems like the "right" way to do this, so if we agree that's the case then perhaps then we don't care about adding a little bit of additional code.

Any thoughts here @qualidafial ?

**Comment by qualidafial** [comment]
What about just adding a type to the `@BindBean` annotation?

    selectStuff(@BindBean(type=PublicInterface.class) PublicInterface param)

**Comment by gsmith85** [comment]
So this was an option I considered.  We could arugment `BindBean` with a parameter type:

```
public @interface BindBean
{
// ...
`Class<?> parameterType();`
}
```

And have access to the `parameterType` in our `BindBeanFactory`.  Which is great.

However, by doing so, we've just forced everyone who upgrades their version of JDBI2 to the latest to  change their `@BindBean` annotation to include a type token.  Required changes of JDBI consumers aside, it also seems a bit odd to have the API require the specification of a class token in duplicate to the type provided for the parameter itself.  It may have the one upside that for raw type iterables, you could specify a type that you expect all objects to be members of.

Of course we could provide a default for `parameterType` but `null` is not an option per the JLS.  With `null` we'd be able to say "use the type of arg as a default" and otherwise use parameterType ... but without it we'd have ambiguity of whether the parameterType was specified by the user or not.

If we wanted to use two additional optional fields:

```
public @interface BindBean
{
// ...
    boolean useTypeOverride() default false; 
    Class<?> typeOverride() default Object.class;
}
```

We could do `@BindBean(useTypeOverride=true, typeOverride=PublicInterface.class)`.

But then we're back at: should we ever be using `arg`'s type's method be used as the basis of invokation as a default?

TLDR: If we're looking for an extremely lightweight approach and saying "we'll solve this the right way in 3" we could add two fields to `@BindBean`. But my take is the bifurcated approach suggested above handles things the right way for the consumer of JDBI transparently and keeps the API simpler (no additional configuration).

**Comment by qualidafial** [comment]
We already have a proven approach in `@SingleValueResult` which makes the type parameter optional:

    public @interface SingleValueResult
    {
        Class<?> value() default Default.class;
    
        final class Default{}
    }

**Comment by gsmith85** [comment]
Okay, I actually was unaware it was possible to declare an inner class in an `interface`.  So I just learned something new and that simplifies things considerably!

The only question then is whether we're okay using arg's type by default.  It sounds like this is something that will be reconsidered in JDBI3, so if that's the focus right now, then perhaps it makes sense not to address this (and further it's possibly risky in that it could change behavior of `@BindBean` for users who upgrade to the latest 2.X release).

Give me your thoughts and I can get this code wrapped up.

**Comment by qualidafial** [comment]
We've been using arg's runtime type for some time, and this is the first report I've seen of it being a problem.

I think the safer approach is to add the `type()` method to the annotation. This way the new behavior is opt-in and existing users will be unaffected.
